workflows/cloudify_system_workflows/snapshots/credentials.py

```diff
@@ -182,8 +149,38 @@ def _create_agent_key_path_dict(self):
             deployment_id = elem[1]
             node_properties = pickle.loads(elem[2])
             agent_config = self._get_agent_config(node_properties)
+            tenant = elem[3]
             if 'key' in agent_config:
                 agent_key_path = agent_config['key']
                 key = deployment_id + '_' + node_id
-                agent_key_path_dict[key] = os.path.expanduser(agent_key_path)
+                agent_key_path_dict.setdefault(
+                    key, {})[tenant] = os.path.expanduser(agent_key_path)
         return agent_key_path_dict
+
+
+def add_key_secret(tenant, key_path, key_data):
+    key_name = SECRET_STORE_AGENT_KEY_PREFIX + ''.join(
+        char if char in ALLOWED_KEY_CHARS else '_'
+        for char in key_path
+        )
+
+    client = CloudifyClient(
+        'localhost',
+        token=ctx.rest_token,
+        tenant=tenant,
+    )
+    secrets = client.secrets
+
+    while True:
```

Comment:
How would you suggest revising?
We are looping until we find either a matching auth key (line 183) or a secret key name which isn't used yet (this is needed because I'm not allowed to use `/` and a whole bunch of other stuff, so we might get key name collisions for filenames which are distinct)

When I started writing this function I initially had something along the lines of

```python
secret_value = None
while secret_value != key_data and ...:
    ...
```
but I think that's actually much less clear, and the chance of this accidentally being an infinite loop is just the same

(of course I actually *did* forget the bit which mutates the name before we try again, so I do agree it's a concern. Fixed that now :)